### PR TITLE
.Net: [OpenAPI] Support for `form` style query string parameters of array type

### PR DESCRIPTION
**Pull request by SergeyMenshykh**
### Motivation and Context
This change is required to support two ways of serializing array query string parameters of 'form' style:
1. A query string parameter per array item - p1=a&p1=b&p1=c.
2. A comma-separated value per array item - p1=a,b,c.

image

Related issue - https://github.com/microsoft/semantic-kernel/issues/2745

### Description
- The `FormStyleParameterSerializer` class has been added to perform `form` style parameter serialization. Currently, it is only used for query string parameter serialization but can be used later for serializing `cookie` parameters as well. For more details about parameter styles, please refer to the [link](https://swagger.io/specification/v3/#parameter-object).
- The `QueryStringBuilder` class has been modified to use the `FormStyleParameterSerializer` for parameter serialization.
- The `RestApiOperationParameter` class has been extended with the new `Explode` property, which indicates whether arrays and objects should generate separate parameters for each array item.

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/src/Functions/Functions.OpenAPI/Builders/Serialization/FormStyleParameterSerializer.cs [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+            throw new ArgumentNullException(nameof(parameter));
+        }
+
+        if (parameter.Style != RestApiOperationParameterStyle.Form)
+        {
+            throw new SKException($"Unexpected Rest Api operation parameter style - `{parameter.Style}`");
+        }
+
+        // Handling parameters of array type.
+        if (parameter.Type == ArrayType)
```
Comment (markwallace-microsoft) [inline dotnet/src/Functions/Functions.OpenAPI/Builders/Serialization/FormStyleParameterSerializer.cs:38]: Any reason not to use an enum here?

File: dotnet/src/Functions/Functions.OpenAPI/Builders/Serialization/FormStyleParameterSerializer.cs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+    /// <summary>
+    /// Serializes the items of an array as a single parameter with comma-separated values.
+    /// </summary>
+    /// <param name="name">The name of the parameter.</param>
+    /// <param name="array">The array containing the items to be serialized.</param>
+    /// <returns>A string containing the serialized parameter in the format "name=item1,item2,...".</returns>
+    private static string SerializeArrayItemsAsParameterWithCommaSeparatedValues(string name, JsonArray array)
+    {
+        const string ValuesSeparator = ",";
```
Comment (markwallace-microsoft) [inline dotnet/src/Functions/Functions.OpenAPI/Builders/Serialization/FormStyleParameterSerializer.cs:94]: nit: There is a const here but the other method directly uses "&", nice to be consistent

File: dotnet/src/Functions/Functions.OpenAPI/Extensions/RestApiOperationExtensions.cs [hunk 3]
```diff
@@ -164,6 +167,7 @@ private static List<RestApiOperationParameter> GetParametersFromPayloadMetadata(
                     parameterName,
                     property.Type,
                     property.IsRequired,
+                    false,
```
Comment (markwallace-microsoft) [inline dotnet/src/Functions/Functions.OpenAPI/Extensions/RestApiOperationExtensions.cs:170]: nit: name the argument for readability

File: dotnet/src/Functions/Functions.UnitTests/Connectors/WebApi/Rest/RestApiOperationTests.cs [hunk 4]
```diff
@@ -87,13 +87,14 @@ public void ShouldReplacePathParametersByValuesFromArguments()
...
     [Fact]
-    public void ShouldReplacePathParametersByDefaultValues()
+    public void ItShouldReplacePathParametersByDefaultValues()
     {
         // Arrange
         var parameterMetadata = new RestApiOperationParameter(
             "fake-path-parameter",
             "fake_type",
             true,
+            false,
```
Comment (markwallace-microsoft) [inline dotnet/src/Functions/Functions.UnitTests/Connectors/WebApi/Rest/RestApiOperationTests.cs:97]: nit: name arguments for readability and maintainability

File: dotnet/src/Functions/Functions.UnitTests/OpenAPI/Builders/QueryStringBuilderTests.cs [hunk 5]
```diff
@@ -27,13 +26,17 @@ public void ShouldAddQueryStringParametersAndUseValuesFromArguments()
...
+            false,
+            RestApiOperationParameterLocation.Query,
+            RestApiOperationParameterStyle.Form);
 
         var secondParameterMetadata = new RestApiOperationParameter(
             "p2",
             "fake_type",
             true,
-            RestApiOperationParameterLocation.Query);
+            false,
```
Comment (markwallace-microsoft) [inline dotnet/src/Functions/Functions.UnitTests/OpenAPI/Builders/QueryStringBuilderTests.cs:37]: nite: won't keep repeating about named arguments :-)

**Review by markwallace-microsoft — APPROVED** [review]
LGTM just a few minor clean up items

File: dotnet/src/Functions/Functions.OpenAPI/Builders/Serialization/FormStyleParameterSerializer.cs (hunk 1 above)
Comment (SergeyMenshykh) [reply dotnet/src/Functions/Functions.OpenAPI/Builders/Serialization/FormStyleParameterSerializer.cs:38]: No particular reason, it should have been added as many other things, but it was never done. I add it to my TODO list to address it after I'm done with the "array" support for query string parameters.

File: dotnet/src/Skills/Skills.UnitTests/OpenAPI/Builders/Query/QueryStringBuilderTests.cs [hunk 6]
```diff
@@ -0,0 +1,198 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
```
[the remark this answers is not in the document]
Comment (SergeyMenshykh) [reply dotnet/src/Skills/Skills.UnitTests/OpenAPI/Builders/Query/QueryStringBuilderTests.cs]: Thanks, done.

File: dotnet/src/Functions/Functions.OpenAPI/Builders/Serialization/FormStyleParameterSerializer.cs (hunk 2 above)
Comment (SergeyMenshykh) [reply dotnet/src/Functions/Functions.OpenAPI/Builders/Serialization/FormStyleParameterSerializer.cs:94]: Thanks, done.

File: dotnet/src/Functions/Functions.OpenAPI/Extensions/RestApiOperationExtensions.cs (hunk 3 above)
Comment (SergeyMenshykh) [reply dotnet/src/Functions/Functions.OpenAPI/Extensions/RestApiOperationExtensions.cs]: Fixed.

File: dotnet/src/Functions/Functions.UnitTests/OpenAPI/Builders/QueryStringBuilderTests.cs (hunk 5 above)
Comment (SergeyMenshykh) [reply dotnet/src/Functions/Functions.UnitTests/OpenAPI/Builders/QueryStringBuilderTests.cs]: Fixed.

File: dotnet/src/Functions/Functions.UnitTests/Connectors/WebApi/Rest/RestApiOperationTests.cs (hunk 4 above)
Comment (SergeyMenshykh) [reply dotnet/src/Functions/Functions.UnitTests/Connectors/WebApi/Rest/RestApiOperationTests.cs]: Fixed.

File: dotnet/src/Functions/Functions.OpenAPI/Model/RestApiOperationParameter.cs [hunk 7]
```diff
@@ -52,12 +52,18 @@ public sealed class RestApiOperationParameter
     /// </summary>
     public string? DefaultValue { get; }
 
+    /// <summary>
+    /// Specifies whether arrays and objects should generate separate parameters for each array item or object property.
+    /// </summary>
+    public bool Explode { get; }
```
Comment (lemillermicrosoft) [inline dotnet/src/Functions/Functions.OpenAPI/Model/RestApiOperationParameter.cs:58]: nit: 
```suggestion
    public bool Expand { get; }
```

Explode to me reads as related to throwing exceptions.

**Review by lemillermicrosoft — APPROVED** [review]

File: dotnet/src/Functions/Functions.OpenAPI/Model/RestApiOperationParameter.cs (hunk 7 above)
Comment (SergeyMenshykh) [reply dotnet/src/Functions/Functions.OpenAPI/Model/RestApiOperationParameter.cs:58]: Yep, np, I just used the name OpenAPI model uses. Fixed.

**Pull request merged**